File: web/src/api/paper.ts
import { request } from './client'
import type { PortfolioCurve } from './position'

export interface PaperAccount {
  id: number
  account_id: number
  user_id: number
  initial_cash: number
  cash: number
}

export interface PaperHolding {
  id: number
  symbol: string
  market: string
  name: string
  quantity: number
  avg_cost: number
  remaining_cost?: number
  cost_basis_estimated?: boolean
  cost_basis_note?: string
  valuation_unavailable_reason?: string
  price: number
  quote_ok: boolean // 仅取到当前有效（fresh）行情时为 true；否则按成本估值
  cost: number
  market_value: number
  profit_amount: number
  profit_pct: number
  quote_as_of?: string // 行情数据源时刻（含 stale 的最近已知）
  freshness_status?: string // fresh | stale | unknown
  stale_reason?: string
  last_price?: number // 最近已知价（stale 展示用，不参与估值）
}

export interface PaperOverview {
  account: PaperAccount
  holdings: PaperHolding[]
  market_value: number
  total_assets: number
  total_profit: number
  total_profit_pct: number
  realized_pnl: number
  quote_stale_count?: number // 无当前有效行情、按成本估值的持仓数
  valuation_note?: string // 部分估值说明（总资产非全实时市值）
  currency_unavailable_reason?: string // 非空时现金、资产及盈亏汇总不可用。
  realized_unavailable_reason?: string
  valuation_unavailable_reason?: string
}

export interface PaperTrade {
  id: number
  symbol: string
  market: string
  name: string
  side: 'buy' | 'sell' | 'adjust'
  price: number
  quantity: number
  amount: number
  fee: number
  tax: number
  realized_pnl: number
  trade_date: string // 业务发生日；除权补跑时可能早于 created_at
  created_at: string
}

export interface TradeInput {
  symbol: string
  market: string
  name?: string
  side: 'buy' | 'sell'
  price?: number
  quantity: number
}

export function getPaperOverview(accountId?: number) {
  return request<PaperOverview>({ url: '/paper/overview', method: 'get', params: { account_id: accountId } })
}

export function paperTrade(input: TradeInput, accountId?: number) {
  return request<PaperTrade>({ url: '/paper/trade', method: 'post', data: input, params: { account_id: accountId } })
}

export function getPaperTrades(limit = 50, accountId?: number) {
  return request<PaperTrade[]>({ url: '/paper/trades', method: 'get', params: { limit, account_id: accountId } })
}

export function resetPaper(initialCash?: number, accountId?: number) {
  return request<PaperAccount>({ url: '/paper/reset', method: 'post', data: { initial_cash: initialCash }, params: { account_id: accountId } })
}

// B7 模拟盘资产曲线（读每交易日 16:20 落库的快照；partial 点表示当日有标的无有效行情）。
export function getPaperCurve(days = 90, signal?: AbortSignal, accountId?: number) {
  return request<PortfolioCurve>({ url: '/paper/curve', method: 'get', params: { days, account_id: accountId }, signal })
}

File: server/model/paper.go
package model

import "time"

// 模拟交易（纸上交易）：每用户一个虚拟账户，用真实行情成交与估值，练习不承担真实风险。
const (
	PaperSideBuy  = "buy"
	PaperSideSell = "sell"

	PaperDefaultCash = 100000 // 默认初始资金
)

// PaperAccount 用户的模拟账户（每用户一个）。
type PaperAccount struct {
	ID          int64     `gorm:"primaryKey" json:"id"`
	UserID      int64     `gorm:"uniqueIndex" json:"user_id"`
	InitialCash float64   `gorm:"type:decimal(20,2)" json:"initial_cash"`
	Cash        float64   `gorm:"type:decimal(20,2)" json:"cash"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// PaperHolding 模拟账户内的聚合持仓（每 user+symbol+market 一条）。
// AvgCost 为含买入手续费的每股成本基，便于卖出时算真实净盈亏。
type PaperHolding struct {
	ID       int64   `gorm:"primaryKey" json:"id"`
	UserID   int64   `gorm:"index;index:idx_ph_uniq,unique" json:"user_id"`
	Symbol   string  `gorm:"size:16;index:idx_ph_uniq,unique" json:"symbol"`
	Market   string  `gorm:"size:8;index:idx_ph_uniq,unique" json:"market"`
	Name     string  `gorm:"size:64" json:"name"`
	Quantity float64 `gorm:"type:decimal(20,4)" json:"quantity"`
	AvgCost  float64 `gorm:"type:decimal(20,4)" json:"avg_cost"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// PaperTrade 模拟成交流水。
type PaperTrade struct {
	ID     int64  `gorm:"primaryKey" json:"id"`
	UserID int64  `gorm:"index:idx_pt_user" json:"user_id"`
	Symbol string `gorm:"size:16" json:"symbol"`
	Market string `gorm:"size:8" json:"market"`
	Name   string `gorm:"size:64" json:"name"`
	Side   string `gorm:"size:8" json:"side"` // buy / sell

	Price       float64 `gorm:"type:decimal(20,4)" json:"price"`
	Quantity    float64 `gorm:"type:decimal(20,4)" json:"quantity"`
	Amount      float64 `gorm:"type:decimal(20,2)" json:"amount"` // price*qty
	Fee         float64 `gorm:"type:decimal(20,2)" json:"fee"`
	Tax         float64 `gorm:"type:decimal(20,2)" json:"tax"`
	RealizedPnl float64 `gorm:"type:decimal(20,2)" json:"realized_pnl"` // 卖出时的已实现盈亏（净）

	CreatedAt time.Time `json:"created_at"`
}
